In [1]:
import pandas as pd
import joblib

model = joblib.load("../models/best_model.pkl")
seg_df = pd.read_csv("../data/processed/churn_risk_segmented.csv")


In [2]:
sim_contract  = seg_df.copy()

In [3]:
if "Contract_Month-to-month" in sim_contract.columns:
    sim_contract["Contract_Month-to-month"] = 0

if "Contract_One year" in sim_contract.columns:
    sim_contract["Contract_One year"] = 1


In [ ]:
feature_cols = sim_contract.drop(["churn_probability", "risk_segment"], axis=1).columns
sim_contract["new_prob"] = model.predict_proba(sim_contract[feature_cols])[:, 1]


In [5]:
original_mean = seg_df["churn_probability"].mean()
new_mean = sim_contract["new_prob"].mean()
improvement = (original_mean - new_mean) * 100

print(f"Contract strategy churn reduction: {improvement:.2f}%")


Contract strategy churn reduction: 7.56%


In [6]:
sim_contract.to_csv("../data/processed/scenario_contract.csv", index=False)


In [ ]:
sim_discount = seg_df.copy()
sim_discount["MonthlyCharges"] *= 0.9

feature_cols = sim_discount.drop(["churn_probability","risk_segment"], axis=1).columns
sim_discount["new_prob"] = model.predict_proba(sim_discount[feature_cols])[:, 1]

orig = seg_df["churn_probability"].mean()
new = sim_discount["new_prob"].mean()
print(f"Discount Strategy Reduction: {(orig - new)*100:.2f}%")



Discount Strategy Reduction: 1.04%


In [12]:
sim_discount.to_csv("../data/processed/scenario_discount.csv", index=False)
print("saved scenario_discount.csv")


saved scenario_discount.csv


In [4]:
sim_support = seg_df.copy()
if "TechSupport_No" in sim_support.columns:
    sim_support["TechSupport_No"] = 0
    sim_support["TechSupport_Yes"] = 1

feature_cols = sim_support.drop(["churn_probability","risk_segment"], axis=1).columns
sim_support["new_prob"] = model.predict_proba(sim_support[feature_cols])[:, 1]

orig = seg_df["churn_probability"].mean()
new = sim_support["new_prob"].mean()
print(f"Tech Support Strategy Reduction: {(orig - new)*100:.2f}%")





Tech Support Strategy Reduction: -0.00%


In [5]:
sim_support.to_csv("../data/processed/scenario_support.csv", index=False)
print("saved scenario_support.csv")


saved scenario_support.csv


## 📅 Retention Strategy Simulation Summary

Today we simulated potential business interventions to reduce customer churn. 
Instead of only predicting churn, we evaluated how different strategies might 
change churn probability based on our trained model.

## 🧠 Strategies Simulated:

**1. Contract Upgrade Strategy**
- Converts "Month-to-Month" customers to "One-Year"
- Result: Lower churn probability due to reduced contract flexibility

**2. Discount Pricing Strategy**
- Applied a 10% reduction on `MonthlyCharges`
- Result: Pricing incentives can reduce churn among price-sensitive customers

**3. Tech Support Add-On Strategy**
- Adds Tech Support service for customers without it
- Result: Value-added support improves customer stickiness

## 📊 Key Insight

These simulations provide a view into:
- Which levers have the strongest impact on churn
- Estimated churn reduction per strategy
- Potential business actions for retention

## 📁 Artifacts Exported for Dashboard

The following files were generated:

- `scenario_contract.csv`
- `scenario_discount.csv`
- `scenario_addons.csv`

These will be used in future steps for:
- Power BI dashboards
- Retention analysis
- Customer segmentation storytelling

---

This elevates your project from **“ML model”** → **“business decision tool”**.

---

## 💡 Estimated Results Section

- Contract strategy reduction: 7.56%
- Discount Strategy Reduction: 1.04%
- Tech Support Strategy Reduction: -0.00%

